dask/layers.py

```diff
@@ -1205,18 +1206,36 @@ def __init__(
             annotations=annotations,
         )
 
+    @property
+    def columns(self):
+        """Current column projection for this layer"""
+        return self._columns
+
     def project_columns(self, columns):
         """Produce a column projection for this IO layer.
         Given a list of required output columns, this method
         returns the projected layer.
         """
+        from dask.dataframe.io.utils import DataFrameIOFunction
+
         if columns and (self.columns is None or columns < set(self.columns)):
 
             # Apply column projection in IO function
-            try:
+            if isinstance(self.io_func, DataFrameIOFunction):
                 io_func = self.io_func.project_columns(list(columns))
-            except AttributeError:
-                io_func = self.io_func
+            else:
+                # Allow column projections for other layers, but
+                # raise a FutureWarning
+                try:
+                    io_func = self.io_func.project_columns(list(columns))
+
+                    warnings.warn(
+                        "IO functions must inherit from DataFrameIOFunction "
+                        "to enable column projection in the future.",
```

Comment:
We could remove this restriction if we use the python3.8+ `Protocol`

Reply:
I see - I do think it makes sense for us to define a `DataFrameIOFunction` class. However, by adding a protocol to the mix, we could coninue supporting other functions that happend to define `project_columns` and `columns` attributes. Is that correct?

Reply:
Yes, that's correct. They would not need to explicitly inherit from your `DataFrameIOFunction` class (though they still could, with some caveats)

Reply:
Cool - Just learning about `runtime_checkable`/`Protocol` right now :)

Last commit (19a2614) is testing out the case where we use an explicit `DataFrameIOFunction` class, but decorate it as a `runtime_checkable` `Protocol`. It seems like this allows column projections for non-`DataFrameIOFunction` callables, as long as they define `__call__`, `project_columns` and `columns` attributes (e.g. for [this `DataFrameIOLayer` usage in  geopandas](https://github.com/geopandas/dask-geopandas/blob/72cfc48c44f947c68a93c7bcae5e688dca00eccc/dask_geopandas/io/file.py#L9)).

Reply:
> Cool - Just learning about runtime_checkable/Protocol right now :)

I'd really like to write down a protocol for high level graphs/layers in the medium term (similar to what @douglasdavis is doing for collections in #8674 )

> It seems like this allows column projections for non-DataFrameIOFunction callables, as long as they define __call__, project_columns and columns attributes

Yep, that's right.